sensors/triggers/email/email.go

```diff
@@ -0,0 +1,155 @@
+	smtpPassword, err := common.GetSecretFromVolume(emailTrigger.SMTPPassword)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve the smtp password, %w", err)
+	}
+	emailSvc := notifications.NewEmailService(
+		notifications.EmailOptions{
+			Host:     emailTrigger.Host,
+			Port:     int(emailTrigger.Port),
+			Username: emailTrigger.Username,
+			Password: smtpPassword,
+			From:     emailTrigger.From,
+		},
+	)
+	return &EmailTrigger{
+		Sensor:   sensor,
+		Trigger:  trigger,
+		Logger:   logger.With(logging.LabelTriggerType, apicommon.EmailTrigger),
+		emailSvc: emailSvc,
+	}, nil
+}
+
+// GetTriggerType returns the type of the trigger
+func (t *EmailTrigger) GetTriggerType() apicommon.TriggerType {
+	return apicommon.EmailTrigger
+}
+
+func (t *EmailTrigger) FetchResource(ctx context.Context) (interface{}, error) {
+	return t.Trigger.Template.Email, nil
+}
+
+func (t *EmailTrigger) ApplyResourceParameters(events map[string]*v1alpha1.Event, resource interface{}) (interface{}, error) {
+	resourceBytes, err := json.Marshal(resource)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal the Email trigger resource, %w", err)
+	}
+	parameters := t.Trigger.Template.Email.Parameters
+
+	if parameters != nil {
+		updatedResourceBytes, err := triggers.ApplyParams(resourceBytes, t.Trigger.Template.Email.Parameters, events)
+		if err != nil {
+			return nil, err
+		}
+
+		var st *v1alpha1.EmailTrigger
+		if err := json.Unmarshal(updatedResourceBytes, &st); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal the updated Email trigger resource after applying resource parameters, %w", err)
+		}
+
+		return st, nil
+	}
+
+	return resource, nil
+}
+
+// Execute executes the trigger
+func (t *EmailTrigger) Execute(ctx context.Context, events map[string]*v1alpha1.Event, resource interface{}) (interface{}, error) {
+	t.Logger.Info("executing EmailTrigger")
+	_, ok := resource.(*v1alpha1.EmailTrigger)
+	if !ok {
+		return nil, fmt.Errorf("failed to marshal the Email trigger resource")
+	}
+
+	emailTrigger := t.Trigger.Template.Email
+
+	if emailTrigger.To == "" {
+		return nil, fmt.Errorf("to can't be empty")
+	}
+	// check if to emails are valid
+	validEmail := regexp.MustCompile(`^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$`)
+	if !validEmail.MatchString(emailTrigger.To) {
+		return nil, fmt.Errorf("to emailId can't be invalid %v", emailTrigger.To)
+	}
+	body := emailTrigger.Body
+	if body == "" {
+		return nil, fmt.Errorf("body can't be empty")
+	}
+	subject := emailTrigger.Subject
+	if subject == "" {
+		return nil, fmt.Errorf("subject can't be empty")
+	}
+	t.Logger.Infow("sending emails...", zap.Any("to", emailTrigger.To))
+
+	notification := notifications.Notification{
+		Message: emailTrigger.Body,
+		Email: &notifications.EmailNotification{
+			Subject: emailTrigger.Subject,
+			Body:    emailTrigger.Body,
+		},
+	}
+	destination := notifications.Destination{
+		Service:   "email",
+		Recipient: emailTrigger.To,
```

Comment:
How to send to multiple recipients?

Reply:
I was thinking along the lines that the user can create a mail group at their end and use that.

But if we need to support this out of the box, I think we have to either:

- Modify the notifications-engine to support multiple to emails
maybe something like
 ```go
 to := String.Split(dest.Recipient, ",")
 email := smtp.New(smtp.Options{
 ........
 ........
 }).WithSubject(subject).WithBody(body).To(to[0], to[1:]...)
 ```
 in https://github.com/argoproj/notifications-engine/blob/master/pkg/services/email.go#L77-L84
 This shouldn't break any existing functionality

OR

- Send multiple email notifications with each of the to addresses. But this isn't ideal.

OR

- we have to implement a different email service

@whynowy , Would be great to get your opinion on this. Thanks

Reply:
Can you check how the following is implemented?

https://github.com/argoproj/notifications-engine/blob/master/README.md?plain=1#L73

Reply:
My understanding was that it was sending out multiple emails out with different recipients as mentioned in the third option. As this matches with pattern in other applications like slack where a recipient will be a channel or user.

Hence the suggestion of adding support for comma separated emails for recipients to the email notification service at notification-engine.


But I will take another look at it. Thanks.

Reply:
I would suggest:

1. Make the `To` a `[]string`;
2. Do for loop here to send notifications;
3. If there's a follow up in notification-engine to support multiple recipients, we update it here.

Reply:
Cool, sounds like a good plan. Will do the same. Thanks.